File: BEmu/IntradayBarRequest/RequestIntradayBarElementStringArray.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="IntradayBarRequest/RequestIntradayBarElementStringArray.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.IntradayBarRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class RequestIntradayBarElementStringArray : Element
    {
        private readonly string _elementName;
        private readonly List<string> _values;

        internal RequestIntradayBarElementStringArray(string elementName)
        {
            this._elementName = elementName;
            this._values = new List<string>();
        }

        public override int NumElements { get { return 0; } }
        public override int NumValues { get { return this._values.Count; } }
        public override Name Name { get { return new Name(this._elementName); } }
        public override bool IsArray { get { return true; } }
        public override bool IsComplexType { get { return false; } }
        public override bool IsNull { get { return false; } }

        private readonly static string[] _allowedEventTypes = { "TRADE", "BID", "ASK", "BID_BEST", "ASK_BEST", "BEST_BID", "BEST_ASK", "BID_YIELD", "ASK_YIELD", "MID_PRICE", "AT_TRADE", "SETTLE" };
        internal void AddValue(string value)
        {
            if (RequestIntradayBarElementStringArray._allowedEventTypes.Contains(value, StringComparer.Ordinal)) //The API is case-sensitive
                this._values.Add(value);
            else
                throw new KeyNotFoundException(); //yes, the BloombergAPI actaully throws a KeyNotFoundException if the user requests a field that isn't supported.
        }

        internal List<string> Values { get { return this._values; } }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}{1}[] = {{{2}", tabs, this.Name, Environment.NewLine);

            string secs = string.Join(", ", this._values);
            result.AppendFormat("{0}{1}{2}{3}", tabs, Types.IndentType.TAB, secs, Environment.NewLine);

            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);

            return result;
        }
    }
}

File: BEmu_cpp/src/IntradayBarRequest/RequestIntradayBarElementStringArray.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/IntradayBarRequest/RequestIntradayBarElementStringArray.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include "BloombergTypes/Name.h"
#include "BloombergTypes/ElementPtr.h"
#include "Types/IndentType.h"
#include "IntradayBarRequest/RequestIntradayBarElementStringArray.h"

namespace BEmu
{
	namespace IntradayBarRequest
	{
		RequestIntradayBarElementStringArray::RequestIntradayBarElementStringArray(const std::string& elementName)
		{
			this->_name = elementName;
		}

		RequestIntradayBarElementStringArray::~RequestIntradayBarElementStringArray()
		{
			for(std::vector<std::string*>::const_iterator iter = this->_values.begin(); iter != this->_values.end(); ++iter)
			{
				std::string* value = *iter;
				delete value;
			}
		}

		Name RequestIntradayBarElementStringArray::name() const
		{
			Name result(this->_name.c_str());
			return result;
		}

		size_t RequestIntradayBarElementStringArray::numValues() const
		{
			return this->_values.size();
		}

		size_t RequestIntradayBarElementStringArray::numElements() const
		{
			return 0;
		}

		bool RequestIntradayBarElementStringArray::isArray() const
		{
			return true;
		}

		bool RequestIntradayBarElementStringArray::isComplexType() const
		{
			return false;
		}

		std::vector<std::string*> RequestIntradayBarElementStringArray::values()
		{
			return this->_values;
		}

		void RequestIntradayBarElementStringArray::addValue(const std::string& value)
		{
			if(this->isEventTypeAllowed(value))
			{
				std::string* str = new std::string(value);
				this->_values.push_back(str);
			}
			else
				throw elementPtrEx;
		}

		bool RequestIntradayBarElementStringArray::isEventTypeAllowed(const std::string& value)
		{
			return 
				(value.compare("TRADE") == 0) ||
				(value.compare("BID") == 0) ||
				(value.compare("ASK") == 0) ||
				(value.compare("BID_BEST") == 0) ||
				(value.compare("ASK_BEST") == 0) ||
				(value.compare("BEST_BID") == 0) ||
				(value.compare("BEST_ASK") == 0) ||
				(value.compare("BID_YIELD") == 0) ||
				(value.compare("ASK_YIELD") == 0) ||
				(value.compare("MID_PRICE") == 0) ||
				(value.compare("AT_TRADE") == 0) ||
				(value.compare("SETTLE") == 0);
		}

		std::ostream& RequestIntradayBarElementStringArray::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << this->_name << "[] = {" << std::endl;

			int index = 0, lastIndex = this->_values.size() - 1;
			for(std::vector<std::string*>::const_iterator iter = this->_values.begin(); iter != this->_values.end(); ++iter)
			{
				std::string* value = *iter;

				stream << value;

				if(index < lastIndex)
					stream << ", ";
			}

			stream << tabs << '}' << std::endl;

			return stream;
		}



	}
}